libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1132 @@
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected node to be an array when parsing %s at %s", node_str, filename);
+
+      return NULL;
+    }
+
+  elements = json_array_get_elements (array);
+  actions = g_ptr_array_new_with_free_func ((GDestroyNotify) flatpak_remote_ref_action_unref);
+
+  for (iter = elements; iter != NULL; iter = iter->next)
+    {
+      /* We use local_error here so that we can catch and
+       * add detail on the filename if necessary */
+      g_autoptr(GError) local_error = NULL;
+      FlatpakRemoteRefAction *action = NULL;
+      gboolean is_filtered = FALSE;
+
+      if (!action_node_should_be_filtered_out (iter->data, &is_filtered, &local_error))
+        {
+          if (g_error_matches (local_error,
+                               EOS_UPDATER_ERROR,
+                               EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC))
+            {
+              g_autofree gchar *filename = g_file_get_path (file);
+              g_propagate_prefixed_error (error, local_error, "Error parsing %s: ", filename);
+            }
+
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return NULL;
+        }
+
+      if (is_filtered)
+        continue;
+
+      action = flatpak_remote_ref_action_from_json_node (iter->data, &local_error);
+
+      if (action == NULL)
+        {
+          if (g_error_matches (local_error,
+                               EOS_UPDATER_ERROR,
+                               EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC))
+            {
+              g_autofree gchar *filename = g_file_get_path (file);
+              g_propagate_prefixed_error (error, local_error, "Error parsing %s: ", filename);
+              return NULL;
+            }
+          else if (g_error_matches (local_error,
+                                    EOS_UPDATER_ERROR,
+                                    EOS_UPDATER_ERROR_UNKNOWN_ENTRY_IN_AUTOINSTALL_SPEC))
+            {
+              g_autofree gchar *filename = g_file_get_path (file);
+              g_warning ("%s while parsing %s. Skipping this action and it will not be reapplied later, system may be in an inconsistent state from this point forward", local_error->message, filename);
+              g_clear_error (&local_error);
+              continue;
+            }
+
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return NULL;
+        }
+
+      g_ptr_array_add (actions, action);
+    }
+
+  /* Now that we have the remote ref actions, sort them by their ordering */
+  g_ptr_array_sort (actions, sort_flatpak_remote_ref_actions);
+
+  return g_steal_pointer (&actions);
+}
+
+FlatpakRemoteRefActionsFile *
+flatpak_remote_ref_actions_file_new (GPtrArray *remote_ref_actions,
+                                     gint       priority)
+{
+  FlatpakRemoteRefActionsFile *file = g_slice_new0 (FlatpakRemoteRefActionsFile);
+
+  file->remote_ref_actions = remote_ref_actions;
```

Comment:
Since the free function here calls `g_ptr_array_unref()`, assuming that the `FlatpakRemoteRefActionsFile` structure has ownership, you need to acquire a reference here.